### ISSUE TITLE
Automate running of Cypress tests

### ISSUE THREAD

**Issue by dantownsend** [piccolo-orm/piccolo_admin#85]
Now we have some basic Cypress tests, it is worth investigating how we can run these as part of GitHub actions.

Cypress tests can be notoriously slow, so we might not want to run them on every commit - we will have to see what the performance is like.

**Comment by LeMeteore** [comment]
What about something like:
```
on:
    release:
        types:
            - created

jobs:
    cypress-tests:
        name: "Run Cyress tests"
        runs-on: "ubuntu-latest"
        steps:
              run:   # command to run cypress
```

Or something like:
```
on: [push]
jobs:
  cypress-tests:
    name: "Run Cyress tests"
    if: github.actor == 'foobar'
    runs-on: ubuntu-latest
    steps:
      - run: # command to run cypress tests
```

[Doc](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution)

**Comment by dantownsend** [comment]
@LeMeteore Yes, that could work. We need to integrate it with this somehow: https://github.com/cypress-io/github-action

What makes this a bit more challenging is in order to run the Cypress tests, we also need the Python REST API running. There's some info on that [here](https://piccolo-admin.readthedocs.io/en/latest/contributing/index.html). cc @sinisaos who originally wrote the Cypress tests.

**Comment by LeMeteore** [comment]
This if what I've tried so far. 
- [the yaml file](https://github.com/LeMeteore/piccolo_admin/blob/github-action-for-cypress/.github/workflows/cypress.yaml)
- [the error](https://github.com/LeMeteore/piccolo_admin/runs/8074848398?check_suite_focus=true):

```
start server "python -m piccolo_admin.piccolo_admin.example --inflate=10000 & command "python -m piccolo_admin.piccolo_admin.example --inflate=10000 &"
current working directory "/home/runner/work/piccolo_admin/piccolo_admin/admin_ui"
/opt/hostedtoolcache/Python/3.7.13/x64/bin/python -m piccolo_admin.piccolo_admin.example --inflate=10000 &
/opt/hostedtoolcache/Python/3.7.13/x64/bin/python: Error while finding module specification for 'piccolo_admin.piccolo_admin.example' (ModuleNotFoundError: No module named 'piccolo_admin')
/home/runner/work/_actions/cypress-io/github-action/v4/dist/index.js:3582
                error = new Error(`The process '${this.toolPath}' failed with exit code ${this.processExitCode}`);
```
The problem seems to be: I need to be in the admin_ui folder to start the Cypress tests. In the mean time, when I try to launch the Python server, I need to be in the root folder :(

- [how to start a server](https://github.com/cypress-io/github-action#start-server)
- [how to set the working directory](https://github.com/cypress-io/github-action#working-directory)

o/

**Comment by dantownsend** [comment]
@LeMeteore Ah, interesting!

You might be able to do something like this to get the python server working from the `admin_ui` directory:

```
export PYTHONPATH='..' && python -m piccolo_admin.example
```

**Comment by sinisaos** [comment]
@dantownsend Your solution locally works.
I was also able to start the server from `admin_ui` with `python ../piccolo_admin/example.py`.
Cypress tests need to be updated (3 out of 4 don't work, but only small changes need to be made) and I also find that custom forms don't work because `media_columns` is undefined and the form input fields don't render (`Error in render: "TypeError: Cannot read properties of undefined (reading 'media_columns')`).

![custom_forms](https://user-images.githubusercontent.com/30960668/187284317-ec8aec1f-5ae7-4c74-bede-9b61f8fa5d7c.png)

**Comment by dantownsend** [comment]
@sinisaos You're right - there's a bug with custom forms. This is a good example of how the Cypress tests will help us.

I don't think I caught this, because it only happens if you go straight to the custom form (if you click on a table first, the bug doesn't appear).

I'll have a look.

**Comment by sinisaos** [comment]
@dantownsend Yes. Cypress discovered that bug. If @LeMeteore  manages to integrate Cypress into GH actions it will be great. 

> I don't think I caught this, because it only happens if you go straight to the custom form (if you click on a table first, the bug doesn't appear).

You're right. That's strange.

**Comment by dantownsend** [comment]
@sinisaos I think I have a fix - will push it now.

**Comment by sinisaos** [comment]
@dantownsend Everything works now. I will update the Cypress tests and made PR tomorrow.

**Comment by dantownsend** [comment]
@sinisaos Thanks. The custom form test is failing because we made two changes:

 1. The 'Title' input now has a default value of 'Enquiry'
 1. When a form is successfully submitted we now redirect to a success page.

To solve issue 1, we just need to change this in `CustomForm.spec.ts`:

```
cy.get('[name="title"]')
            .clear(). // <- This is new
            .type('Hello')
            .should('have.value', 'Hello')
```

**Comment by sinisaos** [comment]
@dantownsend Yes. I've already changed that and there are a few minor changes to the other tests. Tomorrow I will do PR.

**Comment by LeMeteore** [comment]
I think I have something interesting for you guys :partying_face: 
https://github.com/LeMeteore/piccolo_admin/runs/8093345638?check_suite_focus=true

**Comment by dantownsend** [comment]
@LeMeteore That's great, thanks!

I looked at the output, and the reason the tests are failing is Cypress can't connect to `piccolo_admin/`.

We use session authentication, so have to run the front end and API on the same domain. The way we achieve this during local development is to run Nginx with a [virtualhost file](https://github.com/piccolo-orm/piccolo_admin/blob/master/extra/piccolo_admin) which acts as a proxy.

We [discussed in the past](https://github.com/piccolo-orm/piccolo_admin/issues/86) using Vue's built in support for proxying. This would mean we don't need to run a separate proxy.

Long story short ... the last thing we need to figure out is the proxying:

 1. Somehow run Nginx (sounds difficult)
 2. Run some other proxy
 3. Use Vue's proxying support (as outlined in this [issue](https://github.com/piccolo-orm/piccolo_admin/issues/86))

**Comment by LeMeteore** [comment]
I won't deny, I don't know what I'm doing right now, never used/played with Cypress :) This is where I am right now:
```
  - name: "Install dependencies"
        run: |
          set -ex
          python -m pip install --upgrade pip
          pip install -r requirements/requirements.txt
          pip install -r requirements/test-requirements.txt
          pip install piccolo_admin[faker]
          cd admin_ui/
          npm install
      - name: "NPM Serve"
        run: |
          cd admin_ui
          npm run serve &
          cd ..
      - name: "Run Cypress tests"
        uses: "cypress-io/github-action@v4"
        with:
          browser: chrome
          headed: true
          working-directory: ./admin_ui
          start: bash ../start_server.sh
          config: baseUrl=http://localhost:8000
          ```
          
> but did you maybe try to add config: pageLoadTimeout=100000,baseUrl=http://piccolo admin/ to cypress.yaml to see what happens?

I set the baseUrl to `localhost:8000` because Cypress kept looking for something running at `localhost:3000` :\ 
[Complete file](https://github.com/LeMeteore/piccolo_admin/blob/github-action-for-cypress/.github/workflows/cypress.yaml)

Also, I'm using a [shell script](https://github.com/LeMeteore/piccolo_admin/blob/github-action-for-cypress/start_server.sh) to start the server, writing the commend directly does not seem to work.

> Long story short ... the last thing we need to figure out is the proxying:

Let's suppose we want to take the difficult route of running Nginx :) 
Before starting the Cypress tests, we can ask Github actions to :

- install Nginx
- install an Nginx config file to start a virtualhost
- this virtualhost should proxy pass to localhost:8000
- What else should be done ? 

Let me please read the referenced links.

**Comment by dantownsend** [comment]
> Also, I'm using a [shell script](https://github.com/LeMeteore/piccolo_admin/blob/github-action-for-cypress/start_server.sh) to start the server, writing the commend directly does not seem to work. I guess, I'll have to better rewrite it, then move it somewhere else (scripts folder?)

Yeah, the scripts folder should be fine.

> Let's suppose we want to take the difficult route of running Nginx :)
> Before starting the Cypress tests, we can ask Github actions to :
> 
> install Nginx
> install an Nginx config file to start a virtualhost
> this virtualhost should proxy pass to localhost:8000
> What else should be done ?

I haven't tried getting GitHub actions to run Nginx - would be cool if it worked though.

**Comment by LeMeteore** [comment]
> I haven't tried getting GitHub actions to run Nginx - would be cool if it worked though. 

Spent some hours trying really weird things to make it work lol. I'm progressing though \o/ haha :)
tldr:
```
127.0.0.1    piccolo_admin
127.0.0.1    foobar
127.0.0.1    piccolo_admin
127.0.0.1    foobar
nginx: the configuration file /etc/nginx/nginx.conf syntax is ok
nginx: configuration file /etc/nginx/nginx.conf test is successful
 server {
     listen 80;
     server_name piccolo_admin;

     # By default Nginx only allows us to upload very small files.
     client_max_body_size 20M;

     location / {
         proxy_pass http://127.0.0.1:80;
     }

     location /js {
         proxy_pass http://127.0.0.1:80;
     }

     location /css {
         proxy_pass http://127.0.0.1:80;
     }

     location /sockjs-node {
         proxy_pass http://127.0.0.1:80;
         proxy_http_version 1.1;
         proxy_set_header Upgrade $http_upgrade;
         proxy_set_header Connection 'Upgrade';
     }

     location /api {
         proxy_pass http://127.0.0.1:8000;
     }

     location /public {
         proxy_pass http://127.0.0.1:8000;
     }
 }
--2022-09-01 10:37:31--  http://piccolo_admin/
Resolving piccolo_admin (piccolo_admin)... 127.0.0.1
Connecting to piccolo_admin (piccolo_admin)|127.0.0.1|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10918 (11K) [text/html]
Saving to: ‘/dev/null’

     0K ..........                                            100%  432M=0s

2022-09-01 10:37:31 (432 MB/s) - ‘/dev/null’ saved [10918/10918]
```

**Comment by LeMeteore** [comment]
Good \o/
https://github.com/LeMeteore/piccolo_admin/runs/8133622008?check_suite_focus=true

**Comment by dantownsend** [comment]
Cool!

I wonder why it can't find that HTML element 🤔 

I wonder if we can access the screenshots to work out what the cause is?

**Comment by sinisaos** [comment]
@LeMeteore thanks again for your efforts, you managed to make the integration. Tests runs ok locally. I'm not a Cypress expert, but maybe we need to add  wait at the start of the test because maybe the page isn't loaded which could be the problem.

```javascript
it('Fill login form and redirect to admin page and then logout', () => {

        cy.on("uncaught:exception", (err, runnable) => {
            // returning false here prevents Cypress from
            // failing the test
            return false
        })

        cy.wait(5000); // try adding this or longer might help

        // Fill the username
        cy.get('[name="username"]')
            .type('piccolo')
            .should('have.value', 'piccolo');
```

**Comment by LeMeteore** [comment]
> I wonder why it can't find that HTML element 

I honestly don't know :smile: I was thinking about changing the browser or maybe using using more than one browser.

> I wonder if we can access the screenshots to work out what the cause is? 

Access the screenshots ? In headless mode ? :thinking: 

> Thanks again for your efforts,  
It was a pleasure to help. Thanks for your guidance and your moral support.

> you managed to make the integration. Tests runs ok locally
:partying_face: 

> I'm not a Cypress expert, but maybe we need to add wait at the start of the test because maybe the page isn't loaded which could be the problem. 

I will let you guys handle this part :smile:

**Comment by LeMeteore** [comment]
> What about something that may look like the following:
> 
> ```
> on:
>     release:
>         types:
>             - created
> 
> jobs:
>     cypress-tests:
>         name: "Run Cyress tests"
>         runs-on: "ubuntu-latest"
>         steps:
>               run:   # command to run cypress
> ```
> 
> Or something like:
> 
> ```
> on: [push]
> jobs:
>   cypress-tests:
>     name: "Run Cyress tests"
>     if: github.actor == 'foobar'
>     runs-on: ubuntu-latest
>     steps:
>       - run: # command to run cypress tests
> ```
> 
> Proposition made using this [part](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution) of the documentation.
> 
> If you are okay, I can send a PR.

The original idea was to make this job run only under some conditions. What is your final take on this please ?

**Comment by dantownsend** [comment]
> The original idea was to make this job run only under some conditions. What is your final take on this please ?

Ideally we will run it like the other tests (each time a commit is pushed to a pull request).

It depends how long it takes to run though. If > 5 minutes, we might have to reconsider when to run it.

> Or I can send the PR as it is right now, and let you handle the remaining parts.

It would be useful to create a PR so we can see how it all hangs together. It's up to you whether you want to continue experimenting, or we can try and tweak the last few bits. 👍

**Comment by LeMeteore** [comment]
> It's up to you whether you want to continue experimenting

I've added a part to save the png and mp4 artifacts, I've added a timeout for the whole job. I move the shell script to the scripts folder. I sent a PR.

> or we can try and tweak the last few bits. 

Please do :smile: I'm not able to deal with "why the tests are failing". 

It was my pleasure to help, I'll defintely come back from time to time to help here and there (Python parts) if I can.

o/

**Comment by dantownsend** [comment]
Fixed by https://github.com/piccolo-orm/piccolo_admin/pull/215
